### PR TITLE
Add gfa paths while metagraph to GFA conversion

### PR DESCRIPTION
**Pull request by heracle**
Description:
The GFA format contains some P type lines which are defining paths in the GFA graph. These paths are used by some gfa visualisation tools for highlighting the corresponding edges of the graph.
This PR is generating the path fields while converting metagraph to GFA if `--infile-base` flag is set.

GFA standard: http://gfa-spec.github.io/GFA-spec/GFA1.html
GFA visualisation tool: https://rrwick.github.io/Bandage/

Testing workflow:
```
$> cat small.fna 
>NC_001
TGATGTTTGATGTTTGATGTT
>NC_002
AGATGTGCTGAAAGAGTACGT
>NC_003
TGATGTGCTGAAAGAGTACGA
>NC_002
AGATGTGCTGGAAGAGTACGT

$> cat tiny_query.fa 
>something|
AGATGTGCTGGAAGAGTACGT
>something2|
TGATGTGCTGAAAGAGTACGA
>first|
TGATGTTTGATGTTTGATGTT

```

```
$> ./metagraph build -k 5 -o small small.fna
$> ./metagraph annotate -i small.dbg --anno-header -o small_anno small.fna
$> ./metagraph assemble -v small.dbg -o assembled --infile-base tiny_query.fa --unitigs --to-gfa --compacted --annotator small_anno.column.annodbg
$> ./metagraph assemble -v small.dbg -o assembled_ncom --infile-base tiny_query.fa --unitigs --to-gfa --annotator small_anno.column.annodbg
```
cmd full paste: https://ideone.com/Apkd33
`cat assembled.gfa` (compacted) https://ideone.com/Lh078R
`cat assembled_ncom.gfa` (not compacted) https://ideone.com/xFOvYh

Signed-off-by: Radu Muntean <radu.muntean.2@gmail.com>

### CODE REVIEW

File: metagraph/src/cli/config/config.cpp [hunk 1]
```diff
@@ -937,6 +937,7 @@ void Config::print_usage(const std::string &prog_name, IdentityType identity) {
             fprintf(stderr, "\t-p --parallel [INT] \tuse multiple threads for computation [1]\n");
             fprintf(stderr, "\n");
             fprintf(stderr, "\t-a --annotator [STR] \t\t\tannotator to load []\n");
+            fprintf(stderr, "\t-i --infile-base [STR] \tload sequences from FASTA file '<infile-base>.fa' to highlight certain GFA paths []\n");
```
Comment (karasikov) [inline metagraph/src/cli/config/config.cpp:940]: put this line next to line 934 with `--to-gfa`

**Review by karasikov — CHANGES_REQUESTED** [review]

**Comment by heracle** [comment]
Thank you very much for the review!

File: metagraph/src/cli/assemble.cpp [hunk 2]
```diff
@@ -174,11 +222,38 @@ int assemble(Config *config) {
                     });
                 }
                 std::lock_guard<std::mutex> lock(str_mutex);
+                is_unitig_end_node.insert(path.back());
```
[the remark this answers is not in the document]
Comment (heracle) [reply metagraph/src/cli/assemble.cpp:224]: Very useful advice, thanks! Changed the set to "tsl/ordered_set"

File: metagraph/src/cli/assemble.cpp [hunk 3]
```diff
@@ -107,6 +108,52 @@ mask_graph(const AnnotatedDBG &anno_graph, Config *config) {
 }
 
 
+std::string sequence_to_gfa_path(std::string seq,
```
[the remark this answers is not in the document]
Comment (heracle) [reply metagraph/src/cli/assemble.cpp]: I missed it. Everything is now a cst reference

File: metagraph/src/cli/assemble.cpp [hunk 4]
```diff
@@ -107,6 +108,52 @@ mask_graph(const AnnotatedDBG &anno_graph, Config *config) {
...
+                                 size_t seq_id,
+                                 std::shared_ptr<DeBruijnGraph> graph,
+                                 std::unordered_set<uint64_t> is_unitig_end_node,
```
[the remark this answers is not in the document]
Comment (heracle) [reply metagraph/src/cli/assemble.cpp]: Changed the set to "tsl/ordered_set". Thanks for the article, it is really interesting

**Review by karasikov — CHANGES_REQUESTED** [review]
It would be great to add some integration tests as well.
Maybe to https://github.com/ratschlab/metagraph/blob/master/metagraph/integration_tests/test_build.py or add a new file `test_gfa.py`

File: metagraph/src/cli/assemble.cpp [hunk 5]
```diff
@@ -107,6 +108,40 @@ mask_graph(const AnnotatedDBG &anno_graph, Config *config) {
...
+                                 const std::shared_ptr<DeBruijnGraph> &graph,
+                                 const tsl::ordered_set<uint64_t> &is_unitig_end_node,
+                                 const Config *config) {
+    std::vector<uint64_t> path_nodes;
+    graph->map_to_nodes_sequentially(seq, [&](uint64_t node) {
+        path_nodes.push_back(node);
+    });
+
+    //    The first node in the path must be a unitig end node.
+    assert(is_unitig_end_node.count(path_nodes[0]));
```
Comment (heracle) [inline metagraph/src/cli/assemble.cpp:122]: Are the PR tests running the integration tests in a different way? Maybe different flags? In the git checks this assert is failing, but while running locally everything is good (`python3 /home/radu/projects/projects2014-metagenome/metagraph/integration_tests/main.py --test_filter "*assemble_gfa*"`)

Comment (karasikov) [reply metagraph/src/cli/assemble.cpp]: Hm, the CI should run the same command...

Comment (heracle) [reply metagraph/src/cli/assemble.cpp]: It is fixed now. I don't understand why yet, but only the github checks in Debug environment were failing. By building the project locally with "-DCMAKE_BUILD_TYPE=Debug" I was able to find and fix the issue. Thanks for reviewing my PRs!

**Comment by heracle** [comment]
pushed https://github.com/ratschlab/metagraph/pull/256

**Pull request closed without merging**